sdk/typescript/src/utils/publish.ts

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Base64DataBuffer } from '../serialization/base64';
+
+/**
+ * Use the sui Command line tool to verify, compile and compute
+ * the topological ordering of the module dependencies within
+ * the package. This method is only available in NodeJS environment.
+ *
+ * Use this method with the local-txn-data-serializer.
+ *
+ * @param packagePath path to the Move package to be published
+ * @param cliPath path to the sui CLI executable. Follow
+ * {@link https://docs.sui.io/devnet/build/install} to install
+ * @returns an array of compiled modules in bytes
+ */
+export function getCompiledModules(
+  packagePath: String,
+  cliPath: String = 'sui'
+): number[][] {
+  const { execSync } = require('child_process');
```

Comment:
I thought putting the `require` inside a function will not load the module unless the function is called, but apparently it is causing the wallet build to fail. Any thoughts about how to fix this?

```
➜ pnpm run build:dev

> sui-wallet@0.0.6 build:dev /Users/cl/ml/fastnft/wallet
> cross-env NODE_ENV="development" TS_NODE_PROJECT="./configs/ts/tsconfig.webpack.json" webpack --progress

ERROR in ../../sdk/typescript/src/utils/publish.ts 17:25-49
Module not found: Error: Can't resolve 'child_process' in '/Users/cl/ml/fastnft/sdk/typescript/src/utils'
```

cc @pchrysochoidis @Jordan-Mysten @Andrew47

Reply:
Webpack tries to resolve it and because by default the `target` env is the browser it can't find the `child_process` module. Probably guarding the function with something like
```
if (typeof window === 'undefined') {
    ....
    return ...;
}
throw new Error('Only available for nodejs....');
```
should fix it

Reply:
If that does not work, an alternative approach, as webpack is being used, would be to add the following to `sdk/typescript/package.json`:

```
"browser": {
    "child_process": false
}
```

See here: https://stackoverflow.com/questions/54459442/module-not-found-error-cant-resolve-child-process-how-to-fix

Reply:
We should only use ES modules, which means if we want this to be dynamic we should use `import()` to load this module.

This does introduce some bundling implications because it forces bundlers to know about built-in node modules, which isn't great.

I think we should either split this into a separate node-only entrypoint, or separate module.

Reply:
> I think we should either split this into a separate node-only entrypoint, or separate module.

Thanks for the suggestion @Jordan-Mysten ! Do you have any link/examples about adding a node-only entrypoint or a separate module?

Reply:
Yeah so concretely I can imagine this two ways:
- A separate file that can be imported like `import { getCompiledModules } from '@mysten/sui.js/publish';`. This would require tweaking the `package.json` and some of the build scripts. This is probably easier.
- Create a new module like `@mysten/sui-publish` that can be downloaded separately. You could just copy a lot of the existing setup to get that going.

The other option is to leave this as-is, but I'd rather we try to avoid node-specific functionality in our core Sui library, so that it could be used in other runtimes (e.g. deno).

Reply:
I also think you could not include this function at all, and just punt this to the user and give them examples for how to do this. Given how simple this functionality is, I don't think that it's really essential that it's wrapped up by the SDK itself.

Reply:
Yeah so concretely I can imagine this two ways:
- A separate file that can be imported like `import { getCompiledModules } from '@mysten/sui.js/publish';`. This would require tweaking the `package.json` and some of the build scripts. This is probably easier.
- Create a new module like `@mysten/sui-node` that can be downloaded separately. You could just copy a lot of the existing setup to get that going.

The other option is to leave this as-is, but I'd rather we try to avoid node-specific functionality in our core Sui library, so that it could be used in other runtimes (e.g. deno).

---

I also think you could not include this function at all, and just punt this to the user and give them examples for how to do this. Given how simple this functionality is, I don't think that it's really essential that it's wrapped up by the SDK itself.

Reply:
> I also think you could not include this function at all, and just punt this to the user and give them examples for how to do this. Given how simple this functionality is, I don't think that it's really essential that it's wrapped up by the SDK itself.

Agreed. I've reverted the changes and added the example instead